### PR TITLE
[JENKINS-66269] Using console notes in `Cause#getShortDescription` precludes usage of the API

### PR DESCRIPTION
**Pull request by basil**
### Problem

See [JENKINS-66269](https://issues.jenkins.io/browse/JENKINS-66269). After upgrading to jenkinsci/pipeline-build-step-plugin#57, we started seeing HTTP 500 errors in the XML API with the following stack trace being printed to the Jenkins log:

```
WARNING o.e.j.s.h.ContextHandler$Context#log: Error while serving http://example.com/job/test/api/xml
org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 2145; An invalid XML character (Unicode: 0x1b) was found in the element content of the document.
	at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:203)
	at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:177)
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400)
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)
	at com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1473)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2923)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:602)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:505)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:842)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:771)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1213)
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:643)
	at org.dom4j.io.SAXReader.read(SAXReader.java:494)
Caused: org.dom4j.DocumentException: Error on line 1 of document  : An invalid XML character (Unicode: 0x1b) was found in the element content of the document.
	at org.dom4j.io.SAXReader.read(SAXReader.java:511)
	at org.dom4j.io.SAXReader.read(SAXReader.java:408)
	at hudson.model.Api.doXml(Api.java:118)
Caused: java.io.IOException: Failed to do XPath/wrapper handling. Turn on FINER logging to view XML.
	at hudson.model.Api.doXml(Api.java:176)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:396)
```

### Evaluation

The following XML is to blame:

```
<action _class="jenkins.model.InterruptedBuildAction">
  <cause _class="org.jenkinsci.plugins.workflow.support.steps.build.DownstreamFailureCause">
    <shortDescription>[8mha:///…[0mtest #1 completed with status FAILURE (propagate: false to ignore)</shortDescription>
  </cause>
</action>
```

Before jenkinsci/pipeline-build-step-plugin#57, these `shortDescription` messages just contained plain text without any special characters. After jenkinsci/pipeline-build-step-plugin#57, they contain a hyperlink which is encoded as a console note with special characters. These special characters are [illegal in XML 1.0](https://en.wikipedia.org/wiki/Valid_characters_in_XML) (and unescaped by the API), so the XML parser rejects them.

### Solution

I thought about trying to escape the control characters and switch to XML 1.1 (which allows, but discourages, such control characters in escaped form). But after thinking about this more deeply, I concluded that it doesn't make sense to be returning these console notes in the API (be it in XML, Python, or JSON) in the first place. Consumers of the API are expecting plain text and won't be able to parse these encoded console notes. We should just strip them out at the API layer.

### Implementation

I reproduced the problem in a test with a synthetic `Cause.UpstreamCause` that has a hyperlink in its `getDescription` method. I then implemented a custom `ExportInterceptor` based on [`ExportInterceptor#DEFAULT`](https://github.com/stapler/stapler/blob/15e251aeab0447b81e54441ff7f9cfda1826f815/core/src/main/java/org/kohsuke/stapler/export/ExportInterceptor.java#L42-L55) that strips console notes. After plumbing through this new `ExportInterceptor` to the `ExportConfig`, the test started passing.

### Proposed changelog entries

Bug: The Remote Access API (in its XML, JSON/JSONP, and Python flavors) now properly decodes console notes in its output.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by basil** [comment]
Or maybe `getShortDescription` shouldn't have hyperlinks in the first place? In which case we could consider jenkinsci/pipeline-build-step-plugin#62 instead of this PR.

**Comment by basil** [comment]
After thinking about this some more, I think we'd probably want to close this in favor of jenkinsci/pipeline-build-step-plugin#62, but I want to wait and see what others think. If others agree then let's take the other PR and close this one.

**Comment by daniel-beck** [comment]
https://github.com/jenkinsci/pipeline-build-step-plugin/pull/57 should be reverted and instead the `Cause` should have an associated Jelly file with the richer UI representation.

**Comment by daniel-beck** [comment]
FWIW I do not understand why this is a build _cause_ rather than some other action in the first place.

File: test/src/test/java/hudson/model/ApiTest.java [hunk 1]
```diff
@@ -174,6 +175,88 @@ public void escapedParameter() throws Exception {
                 page.getWebResponse().getContentAsString());
     }
 
+    private static class HyperlinkCause extends Cause.UpstreamCause {
```
Comment (daniel-beck) [inline test/src/test/java/hudson/model/ApiTest.java:178]: FWIW it's a `CauseOfInterruption`, not a `Cause`.

**Review by jglick — CHANGES_REQUESTED** [review]
https://github.com/jenkinsci/pipeline-build-step-plugin/pull/57 was incorrect. `getShortDescription()` should be plain text. To add hyperlinks like this, override `print(TaskListener)` and/or have a custom `summary` view. (The latter already exists in this case: https://github.com/jenkinsci/pipeline-build-step-plugin/blob/d9a639e92605fa21c3966da60591eb116f1268ff/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/build/DownstreamFailureCause/summary.jelly#L31)

**Comment by basil** [comment]
> To add hyperlinks like this, override `print(TaskListener)`

Did you read my previous comment, in which I wrote: "I think we'd probably want to close this in favor of jenkinsci/pipeline-build-step-plugin#62"? In that PR, I overrode `print(TaskListener)`.

**Comment by jglick** [comment]
Oh sorry, missed that comment.

**Comment by basil** [comment]
Great. I'm glad we agree on the approach taken in jenkinsci/pipeline-build-step-plugin#62. I'll close this PR in favor of that one.

**Pull request closed without merging**